### PR TITLE
[DPE-3292] test mongos relation

### PR DESCRIPTION
**Pull request by MiaAltieri**
## Issue
Validation of mongos/cluster relation doesn't exist

## Solution
Test that relation does what its supposed to do i.e. user creation / removal

## Files to not review
Feel free to ignore the test charm we copied over to `tests/integration/sharding_tests/application/`, this should be removed in the futre

## Future work
Ideally we can deploy a "Test Charm" for mongos to be hosted on instead of this current charm that we have copied over `tests/integration/sharding_tests/application/`. We can either:
1. test with `data-integrator` when it works with `mongos`
2. publish a [test-charm](https://charmhub.io/mysql-test-app) and deploy it in the tests

### CODE REVIEW

File: tests/integration/sharding_tests/application/charmcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Copyright 2023 Canonical Ltd.
```
Comment (MiaAltieri) [inline tests/integration/sharding_tests/application/charmcraft.yaml:1]: Feel free to ignore the test charm we copied over to `tests/integration/sharding_tests/application/`, this should be removed in the future

File: tests/integration/sharding_tests/test_mongos.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        apps=[CONFIG_SERVER_APP_NAME, SHARD_ONE_APP_NAME, MONGOS_HOST_APP_NAME],
+        idle_period=20,
+        raise_on_blocked=False,  # cluster components are blocked waiting for integration.
+        timeout=TIMEOUT,
+        raise_on_error=False,
+    )
+
+
+@pytest.mark.abort_on_fail
+async def test_form_cluster(ops_test: OpsTest) -> None:
```
Comment (Mehdi-Bendriss) [inline tests/integration/sharding_tests/test_mongos.py]: Could we be more explicit regarding the test name, to reflect that we're interested in user counts?

File: tests/integration/sharding_tests/test_mongos.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        app_name=CONFIG_SERVER_APP_NAME,
+        mongos=True,
+        username=username,
+        password=password,
+    )
+
+    mongos_user_client.admin.command("dbStats")
+
+
+async def test_remove_mongos(ops_test: OpsTest) -> None:
```
Comment (Mehdi-Bendriss) [inline tests/integration/sharding_tests/test_mongos.py]: similar comment on naming

File: tests/integration/sharding_tests/test_mongos.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        ops_test,
+        app_name=CONFIG_SERVER_APP_NAME,
+        mongos=True,
+        username=username,
+        password=password,
+    )
+
+    mongos_user_client.admin.command("dbStats")
+
+
```
Comment (Mehdi-Bendriss) [inline tests/integration/sharding_tests/test_mongos.py:115]: nit: (even though not useful in the last test, but just so it's not forgotten for when you add new tests after this one) 
```suggestion
@pytest.mark.abort_on_fail
```

**Review by Mehdi-Bendriss — COMMENTED** [review]
Thanks Mia! I left a few minor comments - mainly with regards to naming

File: tests/integration/sharding_tests/application/lib/charms/mongos/v0/mongos_client_interface.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        database_name: str,
+        extra_user_roles: str,
+        relation_name: str = MONGOS_RELATION_NAME,
+    ) -> None:
+        """Constructor for MongosRequirer object."""
+        self.relation_name = relation_name
+        self.charm = charm
+
+        if not database_name:
+            database_name = f"{self.charm.app}"
```
Comment (dmitry-ratushnyy) [inline tests/integration/sharding_tests/application/lib/charms/mongos/v0/mongos_client_interface.py:123]: nits: 
maybe we can log here the fact that we are falling back to default db name?

File: tests/integration/sharding_tests/helpers.py [hunk 6]
```diff
@@ -1,36 +1,53 @@
...
 
-async def generate_mongodb_client(ops_test: OpsTest, app_name: str, mongos: bool):
+async def generate_mongodb_client(
+    ops_test: OpsTest, app_name: str, mongos: bool, username: str = None, password: str = None
+):
     """Returns a MongoDB client for mongos/mongod."""
     hosts = [unit.public_address for unit in ops_test.model.applications[app_name].units]
-    password = await get_password(ops_test, app_name=app_name)
+    password = password or await get_password(ops_test, app_name=app_name)
+    username = username or "operator"
```
Comment (dmitry-ratushnyy) [inline tests/integration/sharding_tests/helpers.py]: just curious why not set default `username` in `generate_mongodb_client`

**Review by dmitry-ratushnyy — DISMISSED** [review]

File: tests/integration/sharding_tests/application/lib/charms/mongos/v0/mongos_client_interface.py (hunk 5 above)
Comment (MiaAltieri) [reply tests/integration/sharding_tests/application/lib/charms/mongos/v0/mongos_client_interface.py:123]: Hi not sure if you saw this in the description
> Feel free to ignore the test charm we copied over to tests/integration/sharding_tests/application/, this should be removed in the future

It was a test charm copied over from `mongos`

File: tests/integration/sharding_tests/helpers.py (hunk 6 above)
Comment (MiaAltieri) [reply tests/integration/sharding_tests/helpers.py]: great point! thank you!

**Review by dmitry-ratushnyy — DISMISSED** [review]

**Review by Mehdi-Bendriss — DISMISSED** [review]

**Review by Mehdi-Bendriss — APPROVED** [review]

**Review by dmitry-ratushnyy — APPROVED** [review]

**Pull request merged**